controllers/secret.go

```diff
@@ -1,37 +1,115 @@
 package controllers
 
 import (
+	"context"
+	"strconv"
+	"strings"
+
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
 	clusterv1alpha1 "github.com/redhat-et/ipfs-operator/api/v1alpha1"
 )
 
+const (
+	peerIDPrefix     = "peerID-"
+	privateKeyPrefix = "privateKey-"
+)
+
+// errorFunc Returns a function which returns the provided
+// error when it is called.
+func errorFunc(err error) controllerutil.MutateFn {
+	return func() error {
+		return err
+	}
+}
+
 func (r *IpfsReconciler) secretConfig(
+	ctx context.Context,
 	m *clusterv1alpha1.Ipfs,
 	sec *corev1.Secret,
 	clusterSecret,
 	bootstrapPrivateKey []byte,
 ) (controllerutil.MutateFn, string) {
 	secName := "ipfs-cluster-" + m.Name
-	expected := &corev1.Secret{
+
+	expectedSecret := &corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      secName,
 			Namespace: m.Namespace,
 		},
-		Data: map[string][]byte{
-			"CLUSTER_SECRET":          clusterSecret,
-			"BOOTSTRAP_PEER_PRIV_KEY": bootstrapPrivateKey,
-		},
 	}
-	expected.DeepCopyInto(sec)
+	// find secret
+	err := r.Get(ctx, client.ObjectKeyFromObject(expectedSecret), expectedSecret)
+	if err != nil && !errors.IsNotFound(err) {
+		return errorFunc(err), ""
+	}
+	// initialize the secret, if needed
+	if err != nil && errors.IsNotFound(err) {
+		expectedSecret.Data = make(map[string][]byte, 0)
+		expectedSecret.StringData = make(map[string]string, 0)
+		// secret doesn't exist
+		err = generateNewIdentities(expectedSecret, 0, m.Spec.Replicas)
+		if err != nil {
+			return errorFunc(err), ""
+		}
+		expectedSecret.Data["CLUSTER_SECRET"] = clusterSecret
+		expectedSecret.Data["BOOTSTRAP_PEER_PRIV_KEY"] = bootstrapPrivateKey
+	}
+
+	// secret does exist
+	numIdentities := countIdentities(expectedSecret)
+	if numIdentities != m.Spec.Replicas {
+		// create more identities if needed, otherwise they will be reused
+		// when scaling down and then up again
+		if int(numIdentities) < int(m.Spec.Replicas) {
+			// create more
+			err = generateNewIdentities(expectedSecret, numIdentities, m.Spec.Replicas)
+			if err != nil {
+				return errorFunc(err), ""
+			}
+		}
+	}
+
+	expectedSecret.DeepCopyInto(sec)
 	// FIXME: catch this error before we run the function being returned
-	if err := ctrl.SetControllerReference(m, sec, r.Scheme); err != nil {
-		return func() error { return err }, ""
+	if err = ctrl.SetControllerReference(m, sec, r.Scheme); err != nil {
+		return errorFunc(err), ""
 	}
 	return func() error {
+		sec.Data = expectedSecret.Data
 		return nil
 	}, secName
 }
+
+// countIdentities Counts the amount of unique peer identities present in the secret.
+func countIdentities(secret *corev1.Secret) int32 {
```

Comment:
It's a bit strange that we have int32. Is m.Spec.Replicas an int32 for a reason?

There is also casting to int above for comparisons: `if int(numIdentities) < int(m.Spec.Replicas) {`, so something here is not adding up. Ideally we should work with `int`s. (generateIdentities is casting those int32 back into ints again later).

Reply:
We can change from `int32` to `uint32` or `uint64`, but I would like to be explicit about what type of `int` we are using. As far as the two `int` castings - I forgot to delete those when I made commits to the code.

Reply:
We don't really care what the compiler uses for these values. There is little value in choosing specifics ints. Might just be making things slower, but mostly raises the question as to why the choice was made. Up to you anyways, not so important.